src/lsstseries/analysis/structure_function/base_calculator.py

```diff
@@ -46,74 +47,46 @@ class StructureFunctionCalculator(ABC):
-        The expected inputs are 2D lists for time, flux and error.
+    def _equally_weight_lightcurves(self, random_generator=None):
+        """This method reduces the number of difference samples for all light
+        curves to prevent a few from dominating the calculation.
+        """
+        self._get_difference_values_per_lightcurve()
 
-        For each array in time, flux and error, we'll mask out nan's, then use
-        the resulting values to calculate the difference of all pairs.
+        # if the user defined number_lightcurve_samples in the argument container,
+        # use that, otherwise, default to the minimum number of difference values.
+        least_lightcurve_differences = self._argument_container.number_lightcurve_samples or min(
+            self._difference_values_per_lightcurve
+        )
 
-        Then we filter out any differences that correspond to a delta_time <= 0.
+        for lc in self._lightcurves:
+            lc.select_difference_samples(least_lightcurve_differences, random_generator=random_generator)
 
-        We append the resulting numpy arrays into a corresponding python list,
-        and make that available for use in additional calculations.
+    def _get_difference_values_per_lightcurve(self):
+        """Retrieves the number of difference values per lightcurve and stores
+        them in an array.
         """
-        for lc_idx in range(len(self._time)):
-            # Get one numpy array from the list of numpy arrays
-            lc_times = self._time[lc_idx]
-            lc_fluxes = self._flux[lc_idx]
-            lc_errors = self._err[lc_idx]
-
-            # mask out any nan values
-            t_mask = np.isnan(lc_times)
-            f_mask = np.isnan(lc_fluxes)
-            e_mask = np.isnan(lc_errors)  # always mask out nan errors?
-            lc_mask = np.logical_or(t_mask, f_mask, e_mask)
-
-            lc_times = lc_times[~lc_mask]
-            lc_fluxes = lc_fluxes[~lc_mask]
-            lc_errors = lc_errors[~lc_mask]
-
-            # d_times = difference of times
-            dt_matrix = lc_times.reshape((1, lc_times.size)) - lc_times.reshape((lc_times.size, 1))
-            d_times = dt_matrix[dt_matrix > 0].flatten()
-            self._dts.append(d_times)
-
-            # d_fluxes = difference of fluxes
-            df_matrix = lc_fluxes.reshape((1, lc_fluxes.size)) - lc_fluxes.reshape((lc_fluxes.size, 1))
-            d_fluxes = df_matrix[dt_matrix > 0].flatten()
-            self._all_d_fluxes.append(d_fluxes)
-
-            # err^2 = sum of squared errors
-            err2_matrix = (
-                lc_errors.reshape((1, lc_errors.size)) ** 2 + lc_errors.reshape((lc_errors.size, 1)) ** 2
-            )
-            err2s = err2_matrix[dt_matrix > 0].flatten()
-            self._sum_error_squared.append(err2s)
+        self._difference_values_per_lightcurve = [lc.number_of_difference_values for lc in self._lightcurves]
```

Comment:
I'm struggling a bit to understand what the dividing line is between analysis done in the StructureFunctionLightCurve object is versus analysis done in the StructureFunction code itself. Is it just that the lightcurve class is handling all the data serving tasks (difference calculation, bootstrapping), and then the analysis function itself is using those to calculate the sf values? Particularly from a user perspective, I'd be worried about if I had my own structure function, what if I needed it to do something differently with the data? Am I then trying to modify the StructureFunctionLightCurve? Part of this may just be me needing to read this a bit more, but figured I'd just flag my ignorance here.

Reply:
I think what you're sensing is justified, the line between the data manipulation that happens in SFLightCurve and the calculator can be hazy. The way I was viewing this was that the SFLightCurve object would contain the primary data, as well as the differences. Additionally, it would contain the data validation (NaN's, Lists have same number of elements etc.) 

I included the subset selection routine in the class as well for what I perceived as efficiency reasons. The class could maintain the full set of differences in memory, and we could select a subset of those multiple times without having to recreate anything. That function could have gone into `calc_sf2` as well, but I figured it made more sense in the SFLightCurve class. 

I viewed the calculator as doing the Structure Function calculations on one or more SFLightCurve objects (and optionally doing the "combine" action as well). 

To your point about a user with their own structure function, the goal would be that they would not have to edit or subclass the SFLightCurve class. A user would have access to the primary data in the SFLightCurve via `self._times self._fluxes and self._errors`. They would also have access to the full set of difference values via `self.sample_d_times self.sample_d_fluxes and self.sample_sum_squared_error`. (If the user does not apply subsampling to the light curve, each of those represent the full set of difference values. 

So if a user wanted to do something differently with the data, they would be able to manipulate the primary input data within the `calculate` method of their `StructureFunctionCalculator` subclass, or they could manipulate the differences as they see fit. 

That was a super wordy response on my part. I'm happy to chat this out in person if that's easier, because if you're getting this feeling that this is the wrong abstraction, then I want to hear more about it :)

Reply:
Thanks for the detailed response Drew, I'm on board with the idea that we have a class for handling details of the lightcurve (e.g. is it valid? Scrub it for NaNs up front, etc.), and I'm leaning more and more into thinking we want a general lightcurve class that does data validation for any arbitrary analysis function. And then inherited classes like SFlightcurve get to do all the extra bits for their science function. 

And I take your point regarding users still having the equivalent amount of information to work with outside of needing to do any modification of these lightcurve classes, that seems reasonable.

My main concern here is just that I want to make sure that any user who wants to adapt/implement their own science code is able to actually build on top of our codebase without excessive confusion. I think documentation is a perfectly good answer to that for where we are right now.

I'll give this another pass today since it's a big PR, but consider this issue I raised addressed :)

Reply:
Added a comment on issue #68 to be sure to describe working with StructureFunctionLightCurve instances.